### ISSUE TITLE
Stop using `createEditorInput` outside of editor service

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#126960]
We want to move away from allowing anyone to create heavy typed editor inputs. One clear method we need to deprecate is `IEditorService.createEditorInput`. The usages we have today need to be changed to untyped inputs.

We seem to be having a few places left:

**Preferences @roblourens** 

![image](https://user-images.githubusercontent.com/900690/123048294-8adfaa80-d3fe-11eb-802c-ee186e9f22e3.png)

**Custom Editors**

![image](https://user-images.githubusercontent.com/900690/123048352-98953000-d3fe-11eb-9f67-56fd954a4f3d.png)

**Text Editors**

![image](https://user-images.githubusercontent.com/900690/123048518-c7aba180-d3fe-11eb-92a1-a62ad7db89e2.png)

We then have quite a few more usages that imho are less critical because they originate from contexts where we actually do want typed inputs (editor input deserialisation). However, we should probably move those into respective components and out of the editor service itself. I will create a follow up issue on that.

**Comment by bpasero** [comment]
Actually in those places where we must have a typed editor input, we should probably change those to call `IEditorOverrideService#resolveEditorOverride` so that all creation calls for editor inputs are in one place.

**Comment by roblourens** [comment]
Will be deleted first thing in debt week when I get rid of the split json editor 😁

**Comment by bpasero** [comment]
That's what I wanted to hear :)

**Comment by bpasero** [comment]
Challenge for me to reduce usages is that the `IEditorResolverService.resolveEditor` method is `async` but I need my inputs sync. So maybe there needs to be a way to get the `createEditorInput` functionality through some kind of service in those specific locations where explicitly needed.

**Comment by roblourens** [comment]
I am trying replacing the split json editor with a simpler split editor, so I want to create a `SideBySideEditorInput` with a couple editor inputs. @lramos15 says there isn't a better way to do this, so hopefully it's ok to leave it for now. It looks like this.

```
	public createSplitJsonEditorInput(configurationTarget: ConfigurationTarget, resource: URI): IEditorInput {
		const editableSettingsEditorInput = this.editorService.createEditorInput({ resource });
		const defaultPreferencesEditorInput = this.instantiationService.createInstance(TextResourceEditorInput, this.getDefaultSettingsResource(configurationTarget), undefined, undefined, undefined, undefined);
		return new SideBySideEditorInput(editableSettingsEditorInput.getName(), undefined, defaultPreferencesEditorInput, editableSettingsEditorInput);
	}
```

**Comment by bpasero** [comment]
Needs https://github.com/microsoft/vscode/issues/127131

**Comment by bpasero** [comment]
@lramos15 the usages in custom editor land is something I do not fully understand:

https://github.com/microsoft/vscode/blob/21887d672ab602c4ab84dd2c2a166fd2509f997a/src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts#L41-L41

https://github.com/microsoft/vscode/blob/debed6faf38857717c137d6105e5dedd61e470db/src/vs/workbench/contrib/customEditor/browser/customEditors.ts#L156-L157

Why would custom editors every create normal file editor inputs? Isn't that a sign something already went wrong along the way of resolving?

Possibly related: https://github.com/microsoft/vscode/issues/132035

**Comment by bpasero** [comment]
One major challenge is (in preferences for example): the resolver asks to return a `IEditorInput`, but when a contribution wants to use `DiffEditorInput` or `SideBySideEditorInput` where one side is from an area the contribution does not own, it needs to still call `createEditorInput`.

I think we would almost need to change these factories to become `async` and then resolvers could actually resolve the one side of the side by side editor properly...

**Comment by lramos15** [comment]
> Why would custom editors every create normal file editor inputs? Isn't that a sign something already went wrong along the way of resolving?

Sorry, I'm not sure about this code. @mjbvz do you know the history behind this? 

> One major challenge is (in preferences for example): the resolver asks to return a `IEditorInput`, but when a contribution wants to use `DiffEditorInput` or `SideBySideEditorInput` where one side is from an area the contribution does not own, it needs to still call `createEditorInput`.

Could an untyped diff / side by side input have different override for each side?

**Comment by lramos15** [comment]
> Besides: nothing really prevents from showing a editor on the left and a different on the right via side by side editor input. As for diff editor, I agree that makes no sense.

Right so it sounds like we only need to fix the side by side case which you can do by passing something like
`{primary: {resource: 'foo', options: {override: 'default'}}, secondary: {resource :'bar', options: {override: 'hex'}}}`

**Comment by bpasero** [comment]
And actually that is what I tried to fix in my PR to support side by side: no registered editor needs to implement the method to create a side by side input, we take care of it.

We cannot really enforce that both sides of an editor in the diff case have the same override ID but maybe we need to document that and/or have a top level override option in the diff case only which is used. This is related to https://github.com/microsoft/vscode/issues/127637

**Comment by bpasero** [comment]
Having an idea: introduce a new service `ITextEditorInputService` (or similar), which is for creating these text editor inputs. This service is then being used in editor service as fallback and used in all places where today we have `createEditorInput`.

This would make it very clear what `createEditorInput` is about: only creates file, untitled or text (readonly) inputs.

**Comment by lramos15** [comment]
Would this not be moving the problem since everyone would just call upon the `ITextEditorInputService` to create an input?

**Comment by bpasero** [comment]
Yeah, as outlined in chat, I will need such a service for my components that provide text file and untitled support anyway because it has some built in caching of editor inputs. 

I would still think that clients can chose to use this service (which is sync) or the resolver (which is async), but preference would be to use the resolver.

My main intent with this change is:
* to free the editor service from implementing the default editor registration
* to make the default editor very clear when you use it ("text editor")

**Comment by bpasero** [comment]
There is now a new `ITextEditorService`, thus the editor service is no longer exposing a `createEditorInput` method. The remaining usages are in custom editor and preferences land:

![image](https://user-images.githubusercontent.com/900690/131954275-d5100125-78c1-49ec-b5b7-58408749d703.png)

I would suggest that respective component owners decide if this is a fair use of the new service or whether the untyped variant is possible in those places.

**Comment by mjbvz** [comment]
Nice work! I believe I've removed the usages from custom editors

**Comment by bpasero** [comment]
@mjbvz thanks 🥂

**Comment by bpasero** [comment]
Settings editor is the last remaining one.

**Comment by roblourens** [comment]
Sorry, I haven't looked at this in so long and reading it again I don't really understand what you want me to do. Is the problem using `createTextEditor` or is it this code that creates an editor input manually?

https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/workbench/services/preferences/browser/preferencesService.ts#L361-L365

**Comment by bpasero** [comment]
No worries, I think the underlying issue at hand is whether settings editor could adopt the editor resolver API from Logan and not use typed editor inputs directly. The idea is that typed editor inputs (anything that extends `EditorInput`) should not be created upfront for passing to `openEditor` but should be created by the editor component itself as part of the opening via the editor resolver API.

For that to work, you need to register your editor to the editor resolver and provide associated factory method. I think it is fine to keep the `createTextEditor` when resolving such an editor and I think this is already implemented fine here:

https://github.com/microsoft/vscode/blob/7c1c5300513617d79e1c66c8d2117a5b95259ffa/src/vs/workbench/contrib/preferences/common/preferencesContribution.ts#L62

However there are still a few (deprecated) usages of `openEditor` with typed inputs here:

https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/workbench/services/preferences/browser/preferencesService.ts#L350

https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/workbench/services/preferences/browser/preferencesService.ts#L358
